### PR TITLE
Improve Zinc logging API

### PR DESCRIPTION
**Pull request by jvican**
This is the follow-up work to https://github.com/sbt/zinc/pull/304 and all the
comments that @stuhood has provided. Thank you for that.

Every commit is rich in details in its message. If you have any doubt about a
change, I invite you to read it.

I'm attaching a screenshot showing all the commits passing in Scala Center's
CI. I had to remove the first of the commits, and therefore all those passing
commits are lost.

### CODE REVIEW

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt%20validator/parambuild/?zinc=pull/325/head&deploy_to_bintray=false">THIS LINK.</A>

**Comment by jvican** [comment]
scalacenter-zinc-loggers

File: internal/zinc-compile-core/src/main/scala/sbt/internal/inc/FilteredReporter.scala [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    maximumErrors: Int,
+    logger: Logger,
+    positionMapper: Position => Position
+) extends LoggedReporter(maximumErrors, logger, positionMapper) {
+  private final def isFiltered(pos: Position, msg: String, severity: Severity): Boolean = {
+    def isFiltered(filters: Seq[Regex], str: String): Boolean =
+      filters.exists(_.findFirstIn(str).isDefined)
+
+    severity != Severity.Error && (
+      (pos.sourceFile.isPresent && isFiltered(fileFilters, pos.sourceFile.get.getPath)) ||
```
Comment (jvican) [inline internal/zinc-compile-core/src/main/scala/sbt/internal/inc/FilteredReporter.scala]: Note that this uses `getPath` (and not `getAbsolutePath`) on purpose.

**Comment by jvican** [comment]
CI is passing, I have requested three reviews. I need two to merge, as usual. I would appreciate feedback.
  
@eed3si9n @dwijnand I have tried to remove the Travis hook with no luck. It still happens to kick in for my PRs. If you happen to know why or know how to remove it, I would appreciate if you do so.

**Comment by dwijnand** [comment]
Disabled Travis CI for this repo.

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt%20validator/parambuild/?zinc=pull/325/head&deploy_to_bintray=false">THIS LINK.</A>

**Comment by jvican** [comment]
It seems Travis CI is still enabled.

**Comment by stuhood** [comment]
Looks good to me! Thanks @jvican .

**Review by eed3si9n — COMMENTED** [review]
Overall it looks good to me.
Could you open a corresponding PR for sbt so either Dbuild or I can validate it.

**Comment by jvican** [comment]
I'm happy to do it but I don't have access to dbuild. Can you give me permissions to do so?

**Comment by eed3si9n** [comment]
Paging @cunei for Dbuild access.
For now, if you have a PR, I can run the validation.

**Comment by cunei** [comment]
@eed3si9n any user can run validation now

**Comment by typesafe-tools** [comment]
A dbuild validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt%20validator/65/console">is in progress...</a>

**Comment by typesafe-tools** [comment]
[//]: # (Refs: 1.0.x-pull/325/head-1.0-1.0-1.0-1.0.x)
dbuild <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt%20validator/65/">has checked</A> the following projects against Scala 2.12:

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | 1.0.x                | sbt/sbt@25d393bd8b34e3bf72d129c2427349c7aa881701
  zinc              | pull/325/head        | sbt/zinc@b6da4d506e6b5860d1598e1413548293b972d17d
  io                | 1.0                  | sbt/io@880de81bcc9bd4308885d2515d0a1b746026b946
  librarymanagement | 1.0                  | sbt/librarymanagement@7e87603b5fd3fa0d0ce38f6e85755d3973ddf114
  util              | 1.0                  | sbt/util@54cf4a43bb560d2e6fdc48fb51a376ca4228da03
  website           | 1.0.x                | sbt/website@5c32a0afcf64dc600ff154f5456026e2a74dc668

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt%20validator/parambuild?sbt=1.0.x&zinc=pull/325/head&io=1.0&librarymanagement=1.0&util=1.0&website=1.0.x&deploy_to_bintray=false">(restart)</A>

**Comment by jvican** [comment]
@cunei Dbuild is not resolving dependencies.
  
@eed3si9n What's dbuild supposed to tell us? It's clear that this PR breaks sbt because the logger interface has changed. Is this something that you want to have for the posterity?

**Comment by cunei** [comment]
At the moment the test is not really run by dbuild; some more work is needed to make it work properly with sbt 1.0.0-M5/6. The test is currently just an 'sbt compile' done by this script: https://github.com/sbt/sbt/blob/1.0.x/sbt-allsources.sh @dwijnand
You can specify in the build form multiple pull requests for the various components, and the script (and in the future dbuild) will check the changes in all the referenced modules together. For example: https://github.com/sbt/sbt/pull/3294#issuecomment-311832168

**Comment by jvican** [comment]
@cunei Thanks for the explanation. My question really is what's the relation between dbuild and getting this PR merged. /cc @eed3si9n

**Comment by eed3si9n** [comment]
@jvican See https://github.com/sbt/util/pull/85#issuecomment-311832175. This allows us to validate multiple PRs together and make sure that in conjunction they are able to compile sbt. This is to make sure that modules changes propagate within short window of time instead of catching breaking changes at the last minute.

**Comment by jvican** [comment]
> This allows us to validate multiple PRs together and make sure that we in conjunction it's able to compile sbt.
  
Gotcha, so this would include sbt PRs too.
  
Can we please proceed with the merge of this PR? I have work on top of this and I need it to be in so that I can continue at a fast pace.

**Comment by eed3si9n** [comment]
> Can we please proceed with the merge of this PR?

As I wrote 22h ago, could you open a corresponding PR for sbt so either Dbuild or I can validate it?

**Comment by jvican** [comment]
> As I wrote 22h ago, could you open a corresponding PR for sbt so either Dbuild or I can validate it?
  
Didn't get what you were asking me. So you are asking me to create a PR to sbt migrating to my new Zinc changes. Okay.

**Comment by eed3si9n** [comment]
Yea. FWIW I asked Martin to do the same for his watcher stuff too https://github.com/sbt/sbt/pull/3295

**Comment by jvican** [comment]
@cunei Is dbuild also doing `Test/compile`?

**Comment by cunei** [comment]
As I wrote, it's not really dbuild running in the validator, right now. At the moment the validator script it is only running 'sbt compile';my understanding is that 'sbt test', in that script, would not run the scripted tests. Once the real dbuild is working ok with 1.0.0-M6, validation will also run all the tests of the various modules.

**Comment by jvican** [comment]
> As I wrote, it's not really dbuild running in the validator, right now. At the moment the validator script it is only running 'sbt compile';my understanding is that 'sbt test', in that script, would not run the scripted tests. Once the real dbuild is working ok with 1.0.0-M6, validation will also run all the tests of the various modules.
  
I'm asking if dbuild compiles the tests too (hence my previous reference to `Test/compile`). You said that dbuild is only doing `sbt compile` and you should compile tests too if you're compiling sources.

**Comment by cunei** [comment]
No testing is currently done, just compilation of the main code. The tests are not currently being compiled, I believe.

**Comment by jvican** [comment]
> No testing is currently done, just compilation of the main code. The tests are not currently being compiled, I believe.
  
Can you please compile tests too? Otherwise, we don't have the guarantee that the PR passes at least all the compilation steps.

**Comment by cunei** [comment]
I changed `sbt compile` to `sbt ";compile;test:compile"`

**Comment by dwijnand** [comment]
@cunei you can make that `sbt test` so that in addition to compiling everything you run the unit tests; scripted tests aren't run with `test`, they're run with `scripted`.

**Comment by cunei** [comment]
@dwijnand OK, it is now running just `sbt-allsources.sh test`

File: internal/compiler-bridge/src/main/scala/xsbt/CallbackGlobal.scala [hunk 2]
```diff
@@ -159,7 +159,7 @@ sealed class ZincCompiler(settings: Settings, dreporter: DelegatingReporter, out
   private[xsbt] def logUnreportedWarnings(seq: Seq[(String, List[(Position, String)])]): Unit = {
     val drep = reporter.asInstanceOf[DelegatingReporter]
```
Comment (Duhemm) [inline internal/compiler-bridge/src/main/scala/xsbt/CallbackGlobal.scala]: I think that `drep` is not used anymore.

File: internal/zinc-compile-core/src/main/scala/sbt/internal/inc/FilteredReporter.scala [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+ * This implementation has been adapted from the Pants repository.
+ * @link https://github.com/pantsbuild/pants/blob/master/src/scala/org/pantsbuild/zinc/logging/Reporters.scala#L28
+ *
+ * This reporter may be useful to companies that have domain-specific knowledge
+ * about compile messages that are not relevant and can be filtered out, or users
+ * that hold similar knowledge about the piece of code that they compile.
+ */
+class FilteredReporter(
+    fileFilters: Array[Regex],
+    msgFilters: Array[Regex],
```
Comment (Duhemm) [inline internal/zinc-compile-core/src/main/scala/sbt/internal/inc/FilteredReporter.scala]: How about `msgFilters: Array[String => Boolean]`?

File: internal/zinc-compile-core/src/main/scala/sbt/internal/inc/ReporterManager.scala [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+   */
+  private def fromJavaLogLevel(level: Level): SbtLevel.Value = {
+    level match {
+      case Level.INFO    => SbtLevel.Info
+      case Level.WARNING => SbtLevel.Warn
+      case Level.SEVERE  => SbtLevel.Error
+      case Level.OFF     => sys.error("Level.OFF is not supported. Change the logging level.")
+      case _             => SbtLevel.Debug
+    }
+  }
```
Comment (Duhemm) [inline internal/zinc-compile-core/src/main/scala/sbt/internal/inc/ReporterManager.scala:48]: Why don't we remove `sbt.util.Level` altogether then?

Comment (jvican) [reply internal/zinc-compile-core/src/main/scala/sbt/internal/inc/ReporterManager.scala:48]: I think it's too big of a change, doesn't give us much... historically we've always used sbt's level logging and I think it's good to honor that tradition. This is a workaround for a technical Java-Scala compatibility issue.

File: internal/compiler-bridge/src/main/scala/xsbt/CallbackGlobal.scala (hunk 2 above)
Comment (jvican) [reply internal/compiler-bridge/src/main/scala/xsbt/CallbackGlobal.scala]: Good catch, this can be removed, thanks :wink:

**Comment by stuhood** [comment]
Very interested in seeing this one land (and making it into a release). I'm blocked for further testing of zinc 1.0.0-X18+ until it does.

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt%20validator/parambuild/?zinc=pull/325/head&deploy_to_bintray=false">this link.</A>

**Comment by jvican** [comment]
@Duhemm I've addressed all your comments.
  
@jastice You can have a look at this one now, it's been updated. The required changes concern only the compiler logger API in sbt/sbt, changes to get this in should be really small, sbt only needs `ManagedLoggedReporter`.

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt%20validator/parambuild/?zinc=pull/325/head&deploy_to_bintray=false">this link.</A>

**Comment by typesafe-tools** [comment]
A dbuild validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt%20validator/92/console">is in progress...</a>

**Comment by typesafe-tools** [comment]
[//]: # (Refs: pull/3316/head-pull/325/head-1.0-1.0-1.0-1.0.x)
dbuild <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt%20validator/92/">has checked</A> the following projects against Scala 2.12:

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/3316">pull/3316/head      </A> | sbt/sbt@0cc3cfb36444e1efda7c4b5661f8a7ef95fcd7da
  zinc              | <A HREF="https://github.com/sbt/zinc/pull/325">pull/325/head       </A> | sbt/zinc@510d64dd11f0593c592df257bb04cf7c3ee23449
  io                | <A HREF="https://github.com/sbt/io/tree/1.0">1.0                 </A> | sbt/io@f55849d988b94941309d316c43533b0af423f244
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/1.0">1.0                 </A> | sbt/librarymanagement@f8d3718d283cacdc11c6d4a9f101c675f687dc49
  util              | <A HREF="https://github.com/sbt/util/tree/1.0">1.0                 </A> | sbt/util@fe7743878fb590765582aec86d0e45df6ae0a936
  website           | <A HREF="https://github.com/sbt/website/tree/1.0.x">1.0.x               </A> | sbt/website@7e644bc8fc623589152e98bee75afd44893690e3

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt%20validator/parambuild?sbt=pull/3316/head&zinc=pull/325/head&io=1.0&librarymanagement=1.0&util=1.0&website=1.0.x&deploy_to_bintray=false">(restart)</A>

**Comment by typesafe-tools** [comment]
A dbuild validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt%20validator/93/console">is in progress...</a>

**Comment by typesafe-tools** [comment]
[//]: # (Refs: pull/3316/head-pull/325/head-1.0-pull/122/head-1.0-1.0.x)
dbuild <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt%20validator/93/">has checked</A> the following projects against Scala 2.12:

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/3316">pull/3316/head      </A> | sbt/sbt@0cc3cfb36444e1efda7c4b5661f8a7ef95fcd7da
  zinc              | <A HREF="https://github.com/sbt/zinc/pull/325">pull/325/head       </A> | sbt/zinc@510d64dd11f0593c592df257bb04cf7c3ee23449
  io                | <A HREF="https://github.com/sbt/io/tree/1.0">1.0                 </A> | sbt/io@f55849d988b94941309d316c43533b0af423f244
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/pull/122">pull/122/head       </A> | sbt/librarymanagement@e3891c8732dc248bd3ff3be9d5c796454d66bc3a
  util              | <A HREF="https://github.com/sbt/util/tree/1.0">1.0                 </A> | sbt/util@fe7743878fb590765582aec86d0e45df6ae0a936
  website           | <A HREF="https://github.com/sbt/website/tree/1.0.x">1.0.x               </A> | sbt/website@7e644bc8fc623589152e98bee75afd44893690e3

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt%20validator/parambuild?sbt=pull/3316/head&zinc=pull/325/head&io=1.0&librarymanagement=pull/122/head&util=1.0&website=1.0.x&deploy_to_bintray=false">(restart)</A>

**Comment by typesafe-tools** [comment]
A dbuild validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt%20validator/94/console">is in progress...</a>

**Review by eed3si9n — APPROVED** [review]

**Comment by typesafe-tools** [comment]
[//]: # (Refs: pull/3316/head-pull/325/head-1.0-74d8a3835d535d3d787e9294539a0a3efc47674d-1.0-1.0.x)
dbuild <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt%20validator/94/">has checked</A> the following projects against Scala 2.12:

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/3316">pull/3316/head      </A> | sbt/sbt@0cc3cfb36444e1efda7c4b5661f8a7ef95fcd7da
  zinc              | <A HREF="https://github.com/sbt/zinc/pull/325">pull/325/head       </A> | sbt/zinc@510d64dd11f0593c592df257bb04cf7c3ee23449
  io                | <A HREF="https://github.com/sbt/io/tree/1.0">1.0                 </A> | sbt/io@f55849d988b94941309d316c43533b0af423f244
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/74d8a3835d535d3d787e9294539a0a3efc47674d">74d8a3835d535d3d787e9294539a0a3efc47674d</A> | sbt/librarymanagement@74d8a3835d535d3d787e9294539a0a3efc47674d
  util              | <A HREF="https://github.com/sbt/util/tree/1.0">1.0                 </A> | sbt/util@fe7743878fb590765582aec86d0e45df6ae0a936
  website           | <A HREF="https://github.com/sbt/website/tree/1.0.x">1.0.x               </A> | sbt/website@7e644bc8fc623589152e98bee75afd44893690e3

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt%20validator/parambuild?sbt=pull/3316/head&zinc=pull/325/head&io=1.0&librarymanagement=74d8a3835d535d3d787e9294539a0a3efc47674d&util=1.0&website=1.0.x&deploy_to_bintray=false">(restart)</A>

**Pull request merged**